### ISSUE TITLE
[BUG]: can't make dimension constant

### ISSUE THREAD

**Issue by jinbohou714** [MilesCranmer/PySR#957]
### What happened?

import numpy as np
import pandas as pd
from holoviews.plotting.bokeh.styles import font_size
from pysr import PySRRegressor
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
import time
import sympy

# 读取数据
data = pd.read_excel('coefficient_0604.xlsx')
theta = data.iloc[:, 0].values
vx = data.iloc[:, 2].values  # 第三列索引为2
vz = data.iloc[:, 3].values  # 第四列索引为3
TMAC = data.iloc[:, 4].values  # 第五列
TTAC = data.iloc[:, 5].values  # 第六列

# 准备输入特征 (vx, vz)
X = np.column_stack((vx, vz))

# 修改自定义倒数函数为SymPy兼容版本
def sympy_protected_inv(x):
    return sympy.Piecewise(
        (1.0 / x, sympy.Abs(x) > 1e-5),
        (0.0, True)
    )

# 配置PySR模型 - TMAC
print("\n" + "=" * 50)
print("Training PySR for TMAC")
print("=" * 50)
start_time = time.time()

model_tmac = PySRRegressor(
    niterations=500,  # 迭代次数
    populations=50,  # 种群数量
    binary_operators=["+", "-", "*", "/"],  # 二元运算符
    unary_operators=["inv","abs"],  # 简化一元运算符
    extra_sympy_mappings={'inv': sympy_protected_inv},  # 使用SymPy兼容函数

    # 无量纲设置
    dimensionless_constants_only=True,
    dimensional_constraint_penalty=1000.0,

    constraints={
        '/':(0,1),
        'inv': 6,  # 限制inv的使用深度
    },
    maxsize=10,  # 限制表达式最大大小
    parsimony=0.1,  # 增加简洁性因子
    weight_optimize=0.01,  # 优化权重
    progress=True,  # 显示进度条
    procs=4,  # 并行进程数
    batching=False,  # 禁用批处理以避免问题
    warm_start=True,  # 从之前的运行中继续
    random_state=42,  # 随机种子
    model_selection="best",  # 选择最佳模型
    elementwise_loss="loss(x, y) = (x - y)^2",  # 平方损失
)

model_tmac.fit(X, TMAC,
               X_units=["m/s", "m/s"],
               y_units="1")

print(f"Time taken: {time.time() - start_time:.2f} seconds")

# 配置PySR模型 - TTAC
print("\n" + "=" * 50)
print("Training PySR for TTAC")
print("=" * 50)
start_time = time.time()

model_ttac = PySRRegressor(
    niterations=200,
    populations=100,
    binary_operators=['+','-',"*", "/"],
    unary_operators=["inv", "abs"],
    extra_sympy_mappings={'inv': sympy_protected_inv},
    constraints={
        '/':(0,1),
        'inv': 4,
    },
    maxsize=10,  # 限制表达式复杂度
    parsimony=0.1,

    # 无量纲设置
    dimensionless_constants_only=True,
    dimensional_constraint_penalty=1000.0,

    weight_optimize=0.01,
    progress=True,
    procs=4,
    batching=False,
    warm_start=True,
    #random_state=42,
    model_selection="best",
    elementwise_loss="loss(x, y) = (x - y)^2",
)

model_ttac.fit(X, TTAC,
               X_units=["m/s","m/s"],
               y_units = "1")
print(f"Time taken: {time.time() - start_time:.2f} seconds")

# 输出结果 - 直接从模型获取方程
print("\n\n" + "=" * 50)
print("FINAL RESULTS")
print("=" * 50)

def get_best_equation(model):
    """从PySR模型获取最佳方程"""
    try:
        # 尝试获取SymPy表达式
        return model.sympy()
    except Exception as e:
        print(f"Error getting sympy equation: {e}")
        try:
            # 如果失败，尝试获取字符串表达式
            return str(model.get_best()['equation'])
        except:
            try:
                # 如果仍然失败，尝试从模型描述中提取
                return str(model).split("Equation:")[1].split("\n")[0].strip()
            except:
                return "Equation not available"

tmac_eq = get_best_equation(model_tmac)
ttac_eq = get_best_equation(model_ttac)

print(f"TMAC Expression: {tmac_eq}")
print(f"TTAC Expression: {ttac_eq}")

# 计算R²分数
def calculate_r2(model, X, y):
    try:
        pred = model.predict(X)
        return r2_score(y, pred)
    except:
        return float('nan')

tmac_r2 = calculate_r2(model_tmac, X, TMAC)
ttac_r2 = calculate_r2(model_ttac, X, TTAC)

print(f"TMAC Score (R²): {tmac_r2:.4f}")
print(f"TTAC Score (R²): {ttac_r2:.4f}")

# 保存结果到文件
with open('regression_results_pysr.txt', 'w', encoding='utf-8') as f:
    f.write("=" * 50 + "\n")
    f.write("SYMBOLIC REGRESSION RESULTS (PySR)\n")
    f.write("=" * 50 + "\n\n")
    f.write(f"TMAC Expression: {tmac_eq}\n")
    f.write(f"R² Score: {tmac_r2:.4f}\n\n")
    f.write(f"TTAC Expression: {ttac_eq}\n")
    f.write(f"R² Score: {ttac_r2:.4f}\n")

print("Results saved to regression_results_pysr.txt")

# 可视化预测效果
plt.figure(figsize=(12,6))

# TMAC 结果

tmac_pred = model_tmac.predict(X)
ax1 = plt.subplot(1,2,1)
ax1.scatter(TMAC, tmac_pred, alpha=0.6)
ax1.plot([min(TMAC), max(TMAC)], [min(TMAC), max(TMAC)], 'r--')
ax1.set_title('TMAC: Actual vs Predicted',fontsize=12)
ax1.set_xlabel('Actual TMAC',fontsize=12)
ax1.set_ylabel('Predicted TMAC',fontsize=12)

# TTAC 结果

ttac_pred = model_ttac.predict(X)
ax2 = plt.subplot(1,2,2)
ax2.scatter(TTAC, ttac_pred, alpha=0.6, color='orange')
ax2.plot([min(TTAC), max(TTAC)], [min(TTAC), max(TTAC)], 'r--')
ax2.set_title('TTAC: Actual vs Predicted',fontsize=12)
ax2.set_xlabel('Actual TTAC',fontsize=12)
ax2.set_ylabel('Predicted TTAC',fontsize=12)

plt.tight_layout()
plt.savefig('symbolic_regression_results_pysr.png', dpi=300)
plt.show()

# 打印模型方程
print("\nFinal Equations:")
print(f"TMAC = {tmac_eq}")
print(f"TTAC = {ttac_eq}")

### Version

1.5.8

### Package Manager

None

### Interface

Jupyter Notebook

### Relevant log output

```shell

```

### Extra Info

我想回归一个无量纲的数，x1与x2均为有量纲的数，y为无量纲的数，如何使得回归的表达式为无量纲的
